pkg/supportconfig/auth.go

```diff
@@ -0,0 +1,140 @@
+			w.WriteHeader(http.StatusForbidden)
+			_, err = w.Write([]byte(err.Error()))
+			if err != nil {
+				logrus.Errorf("error writing response string in auth %s", err.Error())
+			}
+			return
+		}
+		// if we didn't get any errors while authing, it's safe to call the next function
+		next.ServeHTTP(w, r)
+	})
+}
+
+// authenticate issues a token review for the K8s api - basically you need to have `Authorization: Bearer` set like you
+// would for the K8s api. Right now we only have one endpoint so doing authn and authz at the same time makes sense.
+// however, if this changes, this should be broken out into its own middleware
+func (a *AuthHandler) authenticate(r *http.Request) (*authv1.UserInfo, error) {
+	token, err := getTokenFromHeader(r)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := a.tokenReview.Create(r.Context(), &authv1.TokenReview{
+		Spec: authv1.TokenReviewSpec{
+			Token: token,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// if we aren't authenticated, don't attempt to extract user info
+	if !resp.Status.Authenticated {
+		return nil, fmt.Errorf(unauthenticatedError)
+	}
+	return &resp.Status.User, nil
+}
+
+// authorize checks if a user can get a support config by seeing if they can do "retrieve" on
+// "supportconfigs.managment.cattle.io". As of now, this isn't a real CRD - more a way to do auth in a simple fashion.
+// Intent is that only the admin or a user with these specific perms will be able to get the support configs
+func (a *AuthHandler) authorize(r *http.Request, userInfo *authv1.UserInfo) error {
+	//TODO: Real Resource?
+	resp, err := a.subjectReview.Create(r.Context(), &v1.SubjectAccessReview{
+		Spec: v1.SubjectAccessReviewSpec{
+			ResourceAttributes: &v1.ResourceAttributes{
+				Verb:     "retrieve",
+				Version:  "v1",
+				Resource: "supportconfigs",
+				Group:    "management.cattle.io",
+			},
+			User:   userInfo.Username,
+			Groups: userInfo.Groups,
+			Extra:  toExtra(userInfo.Extra),
+			UID:    userInfo.UID,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	if !resp.Status.Allowed {
+		// if we aren't allowed, return an error with the reason
+		return fmt.Errorf("%s", resp.Status.Reason)
+	}
+	return nil
+}
+
+// getTokenFromHeader finds the converts the header "Authorization: Bearer MyTokenHere" -> MyTokenHere, and returns an
+// error if it fails to do so
+func getTokenFromHeader(r *http.Request) (string, error) {
+	headerVal := r.Header.Get(authHeader)
+	if headerVal == "" {
+		return "", fmt.Errorf("%s not found in headers", authHeader)
+	}
+	headerVal = strings.TrimSpace(headerVal)
+	splitHeader := strings.Split(headerVal, " ")
+	if len(splitHeader) != 2 {
+		return "", fmt.Errorf("header value for %s malformed", authHeader)
+	}
+	// TODO: Check value[0] and ensure it is "Bearer"?
+	return splitHeader[1], nil
+}
+
+// toExtra converts from one type of ExtraValue to another. I dislike this, but it does seem to be necessary
```

Comment:
This is a totally personal nitpicky opinion so feel free to ignore this feedback but I wouldn't include personal commentary in comments like this, it's not necessary and doesn't help someone in the future understand why it's there. If you feel like personal commentary is really important, include your name along with "I" statements because git blame can sometimes get messed up and it's not clear who "I" is who dislikes what's going on here.

fwiw I don't dislike this, it is indeed necessary because Go is a strongly typed language, and we have examples elsewhere doing the same thing https://github.com/rancher/rancher/blob/054b5d195f115b91a5284e85270403ba6f3ac99b/pkg/auth/webhook/webhook.go#L42-L47

Reply:
> This is a totally personal nitpicky opinion so feel free to ignore this feedback but I wouldn't include personal commentary in comments like this, it's not necessary and doesn't help someone in the future understand why it's there.

This is fair and accurate feedback. I will remove the personal commentary. 

> fwiw I don't dislike this, it is indeed necessary because Go is a strongly typed language, and we have examples elsewhere doing the same thing

The thing that I disliked about it (FWIW) is that technically the ExtraValues are both just type aliases to `[]string`. So although they may appear to be a different type, they are really the same type, and I am wasting CPU cycles on converting between two things that are the same. But that's neither here nor there - appears to be necessary.

Reply:
Resolving due to change of goals with this PR.